docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,295 @@
+# ADR 038: KVStore state listening
+
+## Changelog
+
+- 11/23/2020: Initial draft
+
+## Status
+
+Proposed
+
+## Abstract
+
+This ADR defines a set of changes to enable listening to state changes of individual KVStores and exposing these data to consumers.
+
+## Context
+
+Currently, KVStore data can be remotely accessed through [Queries](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/messages-and-queries.md#queries) which proceed through Tendermint and the ABCI.
+In addition to these request/response queries, it would be beneficial to have a means of listening to state changes as they occur in real time.
+
+## Decision
+
+We will modify the `MultiStore` interface and its concrete (`rootmulti` and `cachemulti`) implementations and introduce a new `listenkv.Store` to allow listening to state changes in underlying KVStores.
+We will also introduce two approaches for exposing the data to external consumers: writing to files and writing to a gRPC stream.
+
+### Listening interface
+In a new file- `store/types/listening.go`- we will create a `WriteListener` interface for streaming out state changes from a KVStore.
+
+```go
+// WriteListener interface for writing data out from a listenkv.Store
+type WriteListener interface {
+  // if value is nil then it was deleted
+  OnWrite(key []byte, value []byte)
+}
+```
+
+### Listener type
+We will create a concrete implementation of the `WriteListener` interface in `store/types/listening.go` that gob
+encodes and writes the key value pair into an underlying `io.Writer`.
```

Comment:
Why gob? Why not just length-prefix key/values with their raw bytes since that's what we're proxying anyway? I ask because the true consumer (typically an `io.Writer`) won't necessary be Go-based.

Better to provide a concrete type that doesn't depend on language-specific encoding.

Reply:
Gob was chosen somewhat arbitrarily for its efficiency and simplicity, will adjust this to be language agnostic!

Reply:
We are using protobuf type now